unicode-data-names/lib/Unicode/Char/General/Names.hs

```diff
@@ -23,37 +28,95 @@ module Unicode.Char.General.Names
     ) where
 
 import Control.Applicative ((<|>))
-import Data.Maybe (listToMaybe)
-import Foreign.C.String (CString, peekCAString)
-import System.IO.Unsafe (unsafePerformIO)
+import GHC.Exts
+    ( Addr#, Char(..), Char#, Int#
+    , indexCharOffAddr#, plusAddr#, (+#), (-#), (<#), isTrue#, quotRemInt#
+    , dataToTag#, ord# )
 
+import Unicode.Internal.Bits.Names (unpackNBytes#)
 import qualified Unicode.Internal.Char.UnicodeData.DerivedName as DerivedName
 import qualified Unicode.Internal.Char.UnicodeData.NameAliases as NameAliases
 
 -- | Name of a character, if defined.
 --
 -- @since 0.1.0
-{-# INLINE name #-}
 name :: Char -> Maybe String
-name = fmap unpack . DerivedName.name
+name (C# c#) = case DerivedName.name c# of
```

Comment:
My only concern is that we are making the implementation significantly more complex to make these performance gains. Is the performance gain required for unicode-data-names? Are these APIs to be used at places where the additional performance matters? Maybe that is the reason why ICU performance is not so good, because they do not care about the perf of these APIs.

Reply:
I have the same concern about Text/ByteString APIs. If the performance does not matter just the String APIs are enough. More APIs add significantly more maintenance overhead.

I know I am a bit late about this, you have done a lot of hard work on this. But in general we should optimize only if it matters. I do not have much idea where these APIs are used, so I might be wrong.

Reply:
1. The hard work is already done! It was a good exercise to start with a package with relatively small API. Next come `unicode-data`.
2. _You should not worry about the special cases for the names._ These are specific (huge) ranges of characters related to CJK. Any addition with the same template will be processed automatically, because I designed the generator and the optimisation to require no maintenance effort. Any new template will be treated as standard name; a manual change would be required only if optimisation is worth it; else it will work normally.
3. The implementation is quite similar for the 3 APIs.
4. It is not only about speed perf, but also about library/executable _size_.
5. You suggested from the beginning to use `CString`s. The current code is not a simple lookup any more and requires byte arrays manipulation. So it seems quite natural to use the standard `bytestring` and `text`, doesn’t it?
6. The only maintenance I see is updating at each new Unicode version. The Text and ByteString API rely on internal modules of third-party libs, but these are already considered almost public and seem quite stable.

Reply:
We can also mark the new API as experimental and without guarantee to be maintained.

Reply:
As the author of this package, I would like to have the best performance, but obviously I will not be optimising it forever. This is only the first round after its creation.
I am also the current maintainer, so it is designed with low maintenance effort in mind.

Reply:
Ok. I went through the code only at a high level. Did not dig deep into details. I am assuming your tests will cover any issues.

Would have been nice if we had a single interface which could be adapted to string/bytestring/text. And the users of this library themselves could do it. We are using internals of bytestring/text so whenever they change we will have to adapt to the change. And it is unlikely that users will use these flags. You are enthusiastic and have time at this point, but at some point of time you will get busier and this will become hard for you to maintain. But I will leave this to your judgement, and I know it can be removed if we cannot maintain it.